Refactor keep car awake

THX723 opened the pull request:
New approach to keep car awake without using Sentry Mode for Tesla BLE and Tessie.
Added missing packages required for Tesla BLE.
Added feedback and error handling for TeslaFi (similar to what was done with Tessie).
Updated sample config file for Keep Car Awake section.

THX723 committed:
Update awake_start

THX723 committed:
Update awake_stop

THX723 committed:
Update configure.sh

THX723 committed:
Update teslausb_setup_variables.conf.sample

THX723 committed:
Update teslausb_setup_variables.conf.sample

fixed some indent issues.

THX723 committed:
Update teslausb_setup_variables.conf.sample

another try at indent issues.

marcone commented on pi-gen-sources/00-teslausb-tweaks/files/teslausb_setup_variables.conf.sample:
```diff
@@ -243,107 +243,161 @@ export CAM_SIZE=40G
-# sufficient.
+# In order to keep the car awake, several methods have been implemented:
+# TeslaFi, Tessie, and Bluetooth Low Energy (BLE). Please note that the previous
+# implementation using the (unofficial) Tesla API was deprecated. Tesla had
```
"deprecated" usually means that it still works, but using it is discouraged. In this case, it just doesn't work anymore.

THX723 replied:
revised

marcone commented on pi-gen-sources/00-teslausb-tweaks/files/teslausb_setup_variables.conf.sample:
```diff
@@ -243,107 +243,161 @@ export CAM_SIZE=40G
+#   ++ Direct connection to the car; unlimited.
+#   - Added complexity to configure initially
+#   - Some security vulnerability
+#   - Only supports Pi 4 and Pi Zero 2 W. (Pi Zero W is not supported at this
```
At least one person reported that this works on Rock Pi 4C+, and I suspect that any device with the right bluetooth version will work, so perhaps this should say something about that, rather than list specific devices.

THX723 replied:
revised

marcone commented on pi-gen-sources/00-teslausb-tweaks/files/teslausb_setup_variables.conf.sample:
```diff
@@ -243,107 +243,161 @@ export CAM_SIZE=40G
+#     * Change your TeslaUSB user passwords (pi and root)
+#     * Set user/password for the TeslaUSB AP (if enabled)
+#     * Enable Tesla's pin-to-drive
+#     * If your TeslaUSB has been compromise, immediately change your Tesla.com
```
"... has been compromise**d**"

THX723 replied:
revised

marcone commented on pi-gen-sources/00-teslausb-tweaks/files/teslausb_setup_variables.conf.sample:
```diff
@@ -243,107 +243,161 @@ export CAM_SIZE=40G
+#		       at home. Upon arriving home, Sentry Mode will already be ON
+#		       to power TeslaUSB through archiving. Once archiving completes, a
+#		       command will be sent to turn OFF Sentry Mode, thus allowing the car
+#		       to go back to sleep. Whenever you start a new drive, car will default
```
replace tabs with spaces to reduce indent

THX723 replied:
revised

marcone commented on pi-gen-sources/00-teslausb-tweaks/files/teslausb_setup_variables.conf.sample:
```diff
@@ -243,107 +243,161 @@ export CAM_SIZE=40G
 #
-# Additional Tessie & TeslaFi Requirements:
+#		      Sentry Mode is not used by TeslaUSB. Instead a periodic command is
+#		      sent to the car to keep it awake.
```
replace tabs with spaces to reduce indent

THX723 replied:
revised

marcone commented on pi-gen-sources/00-teslausb-tweaks/files/teslausb_setup_variables.conf.sample:
```diff
@@ -243,107 +243,161 @@ export CAM_SIZE=40G
-# Enable TeslaFi API Command:
-#   Navigate to Settings -> Tesla API -> Commands (tab).
-#   Verify 'Sentry Mode' command is checked (enabled).
+# .   .   .   .   .   .   .   .   .   .   .   .   .   .   .   .   .   .   .   .
```
this seems weird here. These settings are all related so I don't see the need for a separator like this

THX723 replied:
They are steps. In the 'Commands' tabs is a list of supported API commands that can be marked for Enable/Disable.

THX723 replied:
I've added number bullets to minimize confusion.

marcone commented on run/awake_start:
```diff
@@ -2,13 +2,49 @@
+  function nudge {
+    while true
+    do
+      sleep 300
```
in order for the `nudge` function to work, the car will have to keep the USB port powered for an initial 5 minutes without being told to. Is that the case? I thought the USB port was powered off fairly quickly these days. Maybe it would be better to put the sleep at the end of the loop? Also, the old Tesla API "ping" would retry more quickly in case of failure, the idea being that a failure might be due to a transient network problem, and so should be retried again quickly before the car goes to sleep. Maybe worth doing for the Tessie case?

THX723 replied:
It takes~10 minutes after door closing for the power to go out. I actually went back and forth on this one during testing, trying to save 1 extra command used. I supposed it's not enough of a difference with Tessie's limit.
Moved sleep 300 to the top.
Added 1 retry for Tessie.

marcone commented on run/awake_stop:
```diff
@@ -1,52 +1,118 @@
+
+    elif [[ -n "${TESLA_BLE_VIN:+x}" ]]
+    then
+      if [ -e /tmp/keep_awake_task_pid ]
```
If I'm reading the code above correctly this shouldn't happen because the "nudge" background task is only started for SENTRY_CASE=3. It's OK to be thorough, but since this is a "should not happen" kinda thing, maybe at least log it.

THX723 replied:
Oof! That whole thing should not be there. Remnants from my previous implementation before incorporating Tessie support. Gone now.

THX723 committed:
Update teslausb_setup_variables.conf.sample

THX723 committed:
Update teslausb_setup_variables.conf.sample

THX723 committed:
Update awake_start

add retry for Tessie

THX723 committed:
Update awake_start

THX723 committed:
Update awake_start

THX723 committed:
Update awake_start

THX723 committed:
Update awake_stop

THX723 commented:
@marcone All comments addressed.

Scottmg1 commented on pi-gen-sources/00-teslausb-tweaks/files/teslausb_setup_variables.conf.sample:
Minor spelling mistake at Line 391 by the way. Unkown Key should be Unknown Key

THX723 replied:
Thanks Scott. Revised.

THX723 committed:
Update teslausb_setup_variables.conf.sample

THX723 committed:
Update teslausb_setup_variables.conf.sample

THX723 committed:
Update teslausb_setup_variables.conf.sample

miles267 commented:
I just ran the teslausb upgrade script and it removed the required BLE packages that I had to manually install.  Will this address that issue?

THX723 commented:
> I just ran the teslausb upgrade script and it removed the required BLE packages that I had to manually install. Will this address that issue?

Just, as in now-ish? If so, was it from my dev repo/branch?
Also curious which packages did you have to re-install?

I did notice `bluez` & `pi-bluetooth` were missing in the installation process and have included them in this PR.

miles267 commented:
I just re-ran the teslausb upgrade now and saw this during:

The following packages were automatically installed and are no longer required:
  libasound2 libasound2-data
Use 'apt autoremove' to remove them.
The following packages will be REMOVED:
  bluetooth* bluez*
0 upgraded, 0 newly installed, 2 to remove and 0 not upgraded.
After this operation, 3,757 kB disk space will be freed.
(Reading database ... 44379 files and directories currently installed.)
Removing bluetooth (5.50-1.2~deb10u5) ...
Removing bluez (5.50-1.2~deb10u5) ...

After reboot, if I run hcitool dev it says command not found.

When I try to use Tools > Pair BLE it returns:
Failed to send pairing request. Error: ble: failed to enable device: can't init hci: no devices available:

THX723 commented:
> I just re-ran the teslausb upgrade now and saw this during:
> 
> The following packages were automatically installed and are no longer required: libasound2 libasound2-data Use 'apt autoremove' to remove them. The following packages will be REMOVED: bluetooth* bluez* 0 upgraded, 0 newly installed, 2 to remove and 0 not upgraded. After this operation, 3,757 kB disk space will be freed. (Reading database ... 44379 files and directories currently installed.) Removing bluetooth (5.50-1.2~deb10u5) ... Removing bluez (5.50-1.2~deb10u5) ...
> 
> After reboot, if I run hcitool dev it says command not found.
> 
> When I try to use Tools > Pair BLE it returns: Failed to send pairing request. Error: ble: failed to enable device: can't init hci: no devices available:

Which REPO & BRANCH did you install from?

THX723 committed:
Update 00-run.sh

'bluez' is needed for Tesla BLE.

THX723 committed:
Update make-root-fs-readonly.sh

'bluez' is need for Tesla BLE.

miles267 commented:
Whichever is the default in the .conf file.

THX723 committed:
Update teslausb_setup_variables.conf.sample

THX723 commented:
> Whichever is the default in the .conf file.

Ok, then that's expected, as you're not pulling from this PR.
In the meantime, if interested, you may re-install using the following:
```
REPO=THX723
BRANCH=Refactor_keep_car_awake
```

THX723 committed:
Update awake_start

THX723 committed:
Update awake_start

THX723 committed:
Update awake_start

miles267 commented:
> > Whichever is the default in the .conf file.
> 
> Ok, then that's expected, as you're not pulling from this PR. In the meantime, if interested, you may re-install using the following:
> 
> ```
> REPO=THX723
> BRANCH=Refactor_keep_car_awake
> ```

I've just tried this now.  When I update my .conf file and run the teslausb upgrade script, it doesn't show the usual on-screen logging of each step of the update.  Is this because of the REPO and BRANCH change?  It doesn't return any errors and eventually returns to the shell prompt.  Wasn't sure whether that is normal when not using the main branch.

THX723 committed:
Update teslausb_setup_variables.conf.sample

THX723 commented:
> > > Whichever is the default in the .conf file.
> > 
> > 
> > Ok, then that's expected, as you're not pulling from this PR. In the meantime, if interested, you may re-install using the following:
> > ```
> > REPO=THX723
> > BRANCH=Refactor_keep_car_awake
> > ```
> 
> I've just tried this now. When I update my .conf file and run the teslausb upgrade script, it doesn't show the usual on-screen logging of each step of the update. Is this because of the REPO and BRANCH change? It doesn't return any errors and eventually returns to the shell prompt. Wasn't sure whether that is normal when not using the main branch.

That's not normal. I just noticed I fat fingered one of the leading "#" on line 324 of the .conf file. Replace the leading "$" with "#" and rerun /root/bin/setup-teslausb upgrade

miles267 commented:
OK. For the time being I’m using @marcone production branch. And have manually installed blues, bluez-firmware and pi-Bluetooth which I believe have enabled BLE support as expected (I’m able to pair successfully).

What, if anything, does the THX723 repo and refactor car awake part config do differently than I’ve not already enabled via my steps?

THX723 commented:
> OK. For the time being I’m using @marcone production branch. And have manually installed blues, bluez-firmware and pi-Bluetooth which I believe have enabled BLE support as expected (I’m able to pair successfully).
> 
> What, if anything, does the THX723 repo and refactor car awake part config do differently than I’ve not already enabled via my steps?

Keep car awake without involving Sentry Mode.

miles267 commented:
May I ask how it does that? And when it might be migrated into the base code?

THX723 committed:
Update awake_start

Prevent false positives (door already closed, or cable connected) from exiting shell.

THX723 commented:
Through trial and error, I've identified several commands that'll reset the car's sleep timer. Of the bunch, only Closing Charge Port did not have unintended side-effects (from what I can think of).

The PR will merge when macrone have time to do the final review.

marcone commented on run/awake_start:
```diff
@@ -2,13 +2,58 @@
+      # Use Tesla BLE API to nudge the car
+      then
+        log "Telsa BLE: Nudging the car."
+        /root/bin/tesla-control -ble -key-file /root/.ble/key_private.pem -vin "${TESLA_BLE_VIN^^}" charge-port-close || true
```
I assume that since you added that `|| true` to this in the latest commit, that the call to `tesla-control` sometimes fails. Is just ignoring that failure the right approach here? Will the car stay awake another 5 minutes for the next nudge?

THX723 replied:
Yes, correct assumption. Both 'door already closed' and 'cable connected' scenarios are returned as failures. However, the sleep timer still gets reset. I've tested and confirm this extensively using 5 min. (and as much as 10 min. waits).

I gave up trying to make the failure mode handling more robust; filtering out the 2 false negatives from other failures. 

I've tried several ways to capture the error output string without luck. For example:

`BLE_response=$(/root/bin/tesla-control -ble -key-file /root/.ble/key_private.pem -vin "${TESLA_BLE_VIN^^}" charge-port-close)`

I'm not a coder by trade. Maybe you can lend your expertise here.

THX723 replied:
Here's one variation of the test I did.
```
BLE_response=$(/root/bin/tesla-control -ble -key-file /root/.ble/key_private.pem -vin "5YJSA1E68MF453833" charge-port-close)

if [[ "$BLE_response" == *"already closed"* ]]
then
   echo "Door already closed."
elif [[ "$BLE_response" == *"cable connected"* ]]
then
   echo "Charge cable is connected."
else
  #Other Failures
  echo "Tesla BLE: $BLE_response!"
fi
```

THX723 replied:
Either way, it seems as long as the command is interpreted by VCSEC, whether return is success or failure, the sleep timer is reset. So the only scenario that could fail to reset the timer is if there was a comm error; command never received (this is the case I was trying to parse out).

THX723 replied:
Ok, I've figured it out. I need to capture from stderr not stdout. Give me a little time to re-implement.

THX723 replied:
Done

THX723 committed:
Update awake_start

add additional BLE error handling and retries.

THX723 committed:
Update awake_start

fix syntax errors

THX723 committed:
Update awake_start

fix syntax error

THX723 committed:
Update awake_start

delete unintended line

THX723 committed:
Update awake_start

typo

Scottmg1 commented:
Been using this implementation for a bit now @THX723 and it's working great. Thank you.

THX723 commented:
> Been using this implementation for a bit now @THX723 and it's working great. Thank you.

Glad to hear! And thanks for validating it.

THX723 commented:
@marcone Ready for review.

THX723 committed:
Update awake_start

removed VIN placeholder with environment variable

marcone commented on pi-gen-sources/00-teslausb-tweaks/files/teslausb_setup_variables.conf.sample:
```diff
@@ -243,107 +243,164 @@ export CAM_SIZE=40G
+#    4.  If done successfully, you'll see an "Unknown Key" (can be renamed)
+#        entry under the in-car Locks screen.
+#
+#        Note: this does not prevent the car from locking, nor does it allows
```
"... nor does it allow"

THX723 replied:
revised.

THX723 committed:
Update teslausb_setup_variables.conf.sample

grammatical edit

marcone commented on run/awake_start:
```diff
@@ -2,13 +2,75 @@
+      then
+        BLE_response=$(/root/bin/tesla-control -ble -key-file /root/.ble/key_private.pem -vin "${TESLA_BLE_VIN^^}" charge-port-close 2>&1)
+
+        if [[ ( "$BLE_response" == *"already closed"* ) || ( "$BLE_response" == *"cable connected"* ) || ( -z "$BLE_response" ) ]]
```
should this instead/also check the return code of `tesla-control`, e.g. by checking `$?` ?

THX723 replied:
No, since "already closed" and "cable connected" are technically failures and returns 1s (fail).
However, I should change the last condition from a simple -z check to ($? -eq 0).

THX723 committed:
Update awake_start

Changed/add BLE success check condition against ($?).

THX723 committed:
Update awake_start

Previous attempt to add $? to BLE response check exposed some issues with the error handling logic. Rejigged the whole thing and fully validated again with success, false-negatives, and real fault cases.

THX723 commented:
Previous attempt adding $? check to BLE response exposed a flaw in the error handling logic. Rejigged the whole thing and fully validated again with success, false-negatives, and real fault cases. Also reduced the retry wait time to 30s (from 60s). Should still be within Tessie's limit.

Should be good to go now @marcone .

THX723 committed:
Update awake_start

Also added break at the end of retries for Tessie.

miles267 commented:
Looking forward to this being merged into main-dev.

marcone commented on run/awake_start:
```diff
@@ -2,13 +2,82 @@
+          log "Tessie: Failed to send command. Check your TESSIE_API_TOKEN and TESSIE_VIN are set correctly. If they are, your car may be offline or asleep."
+          if [ "$retry" -lt "3" ]
+          then
+            log "Tessie: Retrying again in 60s in case the connection was bad."
```
looks like you changed the sleep to 30 but forgot to update the log message

THX723 replied:
revised.

THX723 committed:
Update awake_start

Revised retry log messages.

THX723 committed:
Merge branch 'main-dev' into Refactor_keep_car_awake

THX723 committed:
Merge branch 'main-dev' into Refactor_keep_car_awake

THX723 commented:
@marcone ready to merge again.

THX723 committed:
Update configure.sh

Fixed typo for TEMPERATURE_INTERVAL check.

THX723 committed:
Revert "Update configure.sh"

This reverts commit fc7b24c3f58ea0aaa5942b91d97fdb277c9b867d.

THX723 commented:
Ignore previous configure.sh commit (now reverted). Was meant for another PR.

THX723 committed:
Merge branch 'main-dev' into Refactor_keep_car_awake

marcone commented on run/awake_start:
```diff
@@ -2,13 +2,82 @@
+      fi
+
+      if [[ -n "${TESLA_BLE_VIN:+x}" ]]
+      #Use Tesla BLE API to nudge the car
```
Nitpick: there are a few places like this where there's no space after the `#`, which is less readable and doesn't match other comments.

THX723 committed:
Update awake_start

Added spaces after comment #s for consistency and legibility.

marcone committed:
Merge branch 'main-dev' into Refactor_keep_car_awake
